src/algebra/direct_limit.lean

```diff
@@ -0,0 +1,483 @@
+/-
+Copyright (c) 2019 Kenny Lau, Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau, Chris Hughes
+
+Direct limit of modules, abelian groups, rings, and fields.
+-/
+
+import linear_algebra.direct_sum_module
+import linear_algebra.linear_combination
+import ring_theory.free_comm_ring
+import ring_theory.ideal_operations
+
+universes u v w u₁
+
+open lattice submodule
+
+variables {R : Type u} [ring R]
+variables {ι : Type v} [nonempty ι]
+variables [directed_order ι] [decidable_eq ι]
+variables (G : ι → Type w) [Π i, decidable_eq (G i)]
+
+class directed_system (f : Π i j, i ≤ j → G i → G j) : Prop :=
+(Hid : ∀ i x h, f i i h x = x)
+(Hcomp : ∀ i j k hij hjk x, f j k hjk (f i j hij x) = f i k (le_trans hij hjk) x)
+
+namespace module
+
+variables [Π i, add_comm_group (G i)] [Π i, module R (G i)]
+
+class directed_system (f : Π i j, i ≤ j → G i →ₗ[R] G j) : Prop :=
+(Hid : ∀ i x h, f i i h x = x)
+(Hcomp : ∀ i j k hij hjk x, f j k hjk (f i j hij x) = f i k (le_trans hij hjk) x)
+
+variables (f : Π i j, i ≤ j → G i →ₗ[R] G j) [directed_system G f]
+
+def direct_limit : Type (max v w) :=
+(span R $ { a | ∃ (i j) (H : i ≤ j) x,
+  direct_sum.lof R ι G i x - direct_sum.lof R ι G j (f i j H x) = a }).quotient
+
+namespace direct_limit
+
+instance : add_comm_group (direct_limit G f) := quotient.add_comm_group _
+instance : module R (direct_limit G f) := quotient.module _
+
+variables (R ι)
+def of (i) : G i →ₗ[R] direct_limit G f :=
+(mkq _).comp $ direct_sum.lof R ι G i
+variables {R ι G f}
+
+theorem of_f {i j hij x} : (of R ι G f j (f i j hij x)) = of R ι G f i x :=
+eq.symm $ (submodule.quotient.eq _).2 $ subset_span ⟨i, j, hij, x, rfl⟩
+
+theorem exists_of (z : direct_limit G f) : ∃ i x, z = of R ι G f i x :=
+nonempty.elim (by apply_instance) $ assume ind : ι,
+quotient.induction_on' z $ λ z, direct_sum.induction_on z
+  ⟨ind, 0, (linear_map.map_zero _).symm⟩
+  (λ i x, ⟨i, x, rfl⟩)
+  (λ p q ⟨i, x, ihx⟩ ⟨j, y, ihy⟩, let ⟨k, hik, hjk⟩ := directed_order.directed i j in
+    ⟨k, f i k hik x + f j k hjk y, by rw [linear_map.map_add, of_f, of_f, ← ihx, ← ihy]; refl⟩)
+
+@[elab_as_eliminator]
+protected theorem induction_on {C : direct_limit G f → Prop} (z : direct_limit G f)
+  (ih : ∀ i x, C (of R ι G f i x)) : C z :=
+let ⟨i, x, h⟩ := exists_of z in h.symm ▸ ih i x
+
+variables {P : Type u₁} [add_comm_group P] [module R P] (g : Π i, G i →ₗ[R] P)
+variables (Hg : ∀ i j hij x, g j (f i j hij x) = g i x)
+include Hg
+
+variables (R ι G f)
+def rec : direct_limit G f →ₗ[R] P :=
```

Comment:
I don't think `rec` is a great name. Usually things like this are called `lift` right?

Reply:
done